pkg/resourcemanager/mysql/mysqluser/mysqluser_reconcile.go

```diff
@@ -0,0 +1,319 @@
+		return false, nil
+	}
+
+	_, err = s.GetDB(ctx, instance.Spec.ResourceGroup, instance.Spec.Server, instance.Spec.DbName)
+	if err != nil {
+		instance.Status.Message = errhelp.StripErrorIDs(err)
+		instance.Status.Provisioning = false
+
+		requeuErrors := []string{
+			errhelp.ResourceNotFound,
+			errhelp.ParentNotFoundErrorCode,
+			errhelp.ResourceGroupNotFoundErrorCode,
+		}
+		azerr := errhelp.NewAzureErrorAzureError(err)
+		if helpers.ContainsString(requeuErrors, azerr.Type) {
+			return false, nil
+		}
+
+		// if the database is busy, requeue
+		errorString := err.Error()
+		if strings.Contains(errorString, "Please retry the connection later") {
+			return false, nil
+		}
+
+		// if this is an unmarshall error - igmore and continue, otherwise report error and requeue
+		if _, ok := err.(*json.UnmarshalTypeError); ok {
+			return false, err
+		}
+	}
+
+	adminUser := string(adminSecret["fullyQualifiedUsername"])
+	adminPassword := string(adminSecret[MSecretPasswordKey])
+	fullServerName := string(adminSecret["fullyQualifiedServerName"])
+
+	db, err := s.ConnectToSqlDb(ctx, MDriverName, fullServerName, instance.Spec.DbName, MSqlServerPort, adminUser, adminPassword)
+	if err != nil {
+		instance.Status.Message = errhelp.StripErrorIDs(err)
+		instance.Status.Provisioning = false
+
+		// catch firewall issue - keep cycling until it clears up
+		if strings.Contains(err.Error(), "is not allowed to connect to this MySQL server") {
+			return false, nil
+		}
+
+		// if the database is busy, requeue
+		errorString := err.Error()
+		if strings.Contains(errorString, "Please retry the connection later") {
+			return false, nil
+		}
+
+		return false, err
+	}
+
+	// determine our key namespace - if we're persisting to kube, we should use the actual instance namespace.
+	// In keyvault we have to avoid collisions with other secrets so we create a custom namespace with the user's parameters
+	key = GetNamespacedName(instance, mysqlUserSecretClient)
+
+	// create or get new user secret
+	DBSecret := s.GetOrPrepareSecret(ctx, instance, mysqlUserSecretClient)
+	// reset user from secret in case it was loaded
+	user := string(DBSecret[MSecretUsernameKey])
+	if user == "" {
+		user = fmt.Sprintf(requestedUsername)
+		DBSecret[MSecretUsernameKey] = []byte(user)
+	}
+
+	// Publishing the user secret:
+	// We do this first so if the keyvault does not have right permissions we will not proceed to creating the user
+	err = mysqlUserSecretClient.Upsert(
+		ctx,
+		key,
+		DBSecret,
+		secrets.WithOwner(instance),
+		secrets.WithScheme(s.Scheme),
+	)
+	if err != nil {
+		instance.Status.Message = "failed to update secret, err: " + err.Error()
+		return false, err
+	}
+
+	user, err = s.CreateUser(ctx, string(instance.Spec.Server), DBSecret, db)
+	if err != nil {
+		instance.Status.Message = "failed creating user, err: " + err.Error()
+		return false, err
+	}
+
+	// apply roles to user
```

Comment:
Is this comment on the wrong line? Should it be for line 165?

Reply:
I think theoretically the line 165 is where the granting of roles happens, but this is above the line where the check of if roles are specified in the spec happens, so should be ok IMO

Reply:
Yeah, so that code chunk is "checking if roles are specified" not "applying roles to user"...
Super minor anyway :)

Reply:
Sure. Update the comments.